### PR TITLE
persist: Add more testing around stats

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR adds some more testing around our `Part` statistics. Specifically it adds two new tests:

1. A `proptest` for correctness. We generate arbitrary `ColumnType`s, and use that `ColumnType` to generate an arbitrary `Vec<Row>`, then we calculate stats on that collection of `Row`s and assert that every `Row` would be included in the stats.
2. A test for stats stability. We use `proptest` with a constant seed to generate 1,000 instances of `RelationDesc`s with at most 4 columns, then a collection of at most 8 `Row`s for these `RelationDesc`s. We generate statistics for all 1,000 scenarios and then take a JSON snapshot of the stats. This test helps us track if any changes occur to our statistics generation.

I'm curious what folks thoughts are on the second test, I'm more than happy to not merge it and use it only to validate https://github.com/MaterializeInc/materialize/pull/27009, if we don't think it provides a ton of signal.

### Motivation

Protect against stats breaking, e.g. in changes like https://github.com/MaterializeInc/materialize/pull/27009

### Tips for reviewer

The PR is broken up into 2 commits:

1. Proptest strategies to generate `Datum`s from a `ColumnType`, and adding the first test.
2. The snapshot test.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

**Comment by ggevay** [comment]
`arb_datum_for_column` could also be used for scalar function property testing! For each scalar function, we could generate some random inputs of the appropriate type using `arb_datum_for_column`, call the function, and check whether `propagates_nulls`, `is_monotone`, and all the other things on `LazyUnaryFunc` are actually true about the results we got. I'll open a follow-up PR for this. I'd be willing to bet a non-trivial amount of money that this will reveal some bugs. E.g., just a few minutes ago [we found](https://github.com/MaterializeInc/materialize/pull/27111#issuecomment-2114745935) some wrong `inverse` annotations on some exotic cast functions. (We already have `test_smoketest_all_builtins`, but that currently uses some hardcoded `interesting_datums` as inputs. Making it use random inputs should be more effective.)

**Comment by danhhz** [comment]
Very cool! @bkirwi mind taking the review on this one?

**Comment by ParkMyCar** [comment]
@ggevay nice! Glad y'all can also find a use for this!

File: src/repr/src/scalar.rs [hunk 1]
```diff
@@ -3767,6 +3796,145 @@ pub fn arb_datum() -> BoxedStrategy<PropDatum> {
...
+        ScalarType::Float64 => any::<f64>().prop_map(PropDatum::Float64).boxed(),
+        ScalarType::Numeric { .. } => arb_numeric().prop_map(PropDatum::Numeric).boxed(),
+        ScalarType::String
+        | ScalarType::PgLegacyName
+        | ScalarType::Char { length: None }
+        | ScalarType::VarChar { max_length: None } => ".*".prop_map(PropDatum::String).boxed(),
+        ScalarType::Char {
+            length: Some(length),
+        } => {
+            // Constrain the length otherwise we try to generate massive strings.
```
Comment (bkirwi) [inline src/repr/src/scalar.rs:3835]: This means we'd generate data that doesn't actually match the type, right? WDYT about capping the length of the `ScalarType::Char`s we generate instead?

**Review by bkirwi — APPROVED** [review]
Nice!

File: src/repr/src/scalar.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/repr/src/scalar.rs]: Doh! yeah that's definitely the way to go, updated!

Also updated the strategy for generating `Char`s to be more realistic

**Pull request merged**